File: 112.path-sum.py
#
# @lc app=leetcode id=112 lang=python3
#
# [112] Path Sum
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if root == None:
            return False
        subtreeTarget = targetSum - root.val
        if root.left == None and root.right == None:
            return subtreeTarget == 0
        elif root.left == None and root.right != None:
            return self.hasPathSum(root.right, subtreeTarget)
        elif root.right == None and root.left != None:
            return self.hasPathSum(root.left, subtreeTarget)
        else:
            return self.hasPathSum(root.left, subtreeTarget) or self.hasPathSum(root.right, subtreeTarget)
        
# @lc code=end

File: 112.path-sum.ts
/*
 * @lc app=leetcode id=112 lang=typescript
 *
 * [112] Path Sum
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
// class TreeNode {
//   val: number;
//   left: TreeNode | null;
//   right: TreeNode | null;
//   constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
//     this.val = val === undefined ? 0 : val;
//     this.left = left === undefined ? null : left;
//     this.right = right === undefined ? null : right;
//   }
// }

function hasPathSum(root: TreeNode | null, targetSum: number): boolean {
  if (root === null) return false;
  const subtreeTargetSum = targetSum - root.val;
  if (root.left === null && root.right === null) {
    return subtreeTargetSum === 0;
  } else if (root.left !== null && root.right === null) {
    return hasPathSum(root.left, subtreeTargetSum);
  } else if (root.left === null && root.right !== null) {
    return hasPathSum(root.right, subtreeTargetSum);
  } else {
    return (
      hasPathSum(root.left, subtreeTargetSum) ||
      hasPathSum(root.right, subtreeTargetSum)
    );
  }
}
// @lc code=end
